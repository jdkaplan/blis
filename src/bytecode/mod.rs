use serde::{Deserialize, Serialize};

pub mod chunk;
pub mod op;

pub use chunk::{Chunk, Constant};
pub use op::{Op, OpError};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Func {
    pub name: String,
    pub arity: u8,
    pub upvalues: u8,
    pub chunk: Chunk,
}
