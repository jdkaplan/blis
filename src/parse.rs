use std::path::Path;

pub type ParseResult<T> = Result<T, (Option<T>, ParseErrors)>;

pub type ParseErrors = Vec<ParseError>;

#[derive(thiserror::Error, Debug)]
#[error("parse error")]
pub struct ParseError;

pub struct File {
    pub name: String,
    pub decls: Vec<Decl>,
}

pub enum Decl {
    Stmt(Stmt),
}

pub enum Stmt {
    Expr(Expr),
}

pub enum Expr {
    Literal(Literal),
}

pub enum Literal {
    U64(u64),
}

pub struct Parser;

impl Parser {
    pub fn parse(path: &Path, source: &str) -> ParseResult<File> {
        let mut errors = Vec::new();
        let mut decls = Vec::new();

        if source.trim() == "1" {
            let decl = Decl::Stmt(Stmt::Expr(Expr::Literal(Literal::U64(1))));
            decls.push(decl);
        } else {
            errors.push(ParseError);
        }

        let file = File {
            name: path.to_string_lossy().into(),
            decls,
        };

        if errors.is_empty() {
            Ok(file)
        } else {
            Err((Some(file), errors))
        }
    }
}
