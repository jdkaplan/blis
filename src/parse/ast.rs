use num_bigint::BigInt;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct Program {
    pub decls: Vec<Declaration>,
}

#[derive(Debug, Clone, Serialize)]
pub enum Declaration {
    Type(Type),
    Func(Func),
    Let(Let),

    Statement(Statement),
}

#[derive(Debug, Clone, Serialize)]
pub enum Statement {
    Break(Break),
    Continue(Continue),
    Loop(Loop),
    Return(Return),

    Assignment(Assignment),
    Expression(Expression),
}

#[derive(Debug, Clone, Serialize)]
pub enum Expression {
    LogicOr(LogicOr),
}

impl Expression {
    pub fn self_terminating(&self) -> bool {
        match self {
            Expression::LogicOr(or) => or.self_terminating(),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct LogicOr {
    pub first: LogicAnd,
    pub rest: Vec<LogicAnd>,
}

impl LogicOr {
    pub fn self_terminating(&self) -> bool {
        self.rest.is_empty() && self.first.self_terminating()
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct LogicAnd {
    pub first: Equality,
    pub rest: Vec<Equality>,
}

impl LogicAnd {
    pub fn self_terminating(&self) -> bool {
        self.rest.is_empty() && self.first.self_terminating()
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Equality {
    Value(Comparison),

    Eq(Box<Equality>, Comparison),
    Ne(Box<Equality>, Comparison),
}

impl Equality {
    pub fn self_terminating(&self) -> bool {
        match self {
            Equality::Value(comp) => comp.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Comparison {
    Value(Term),

    Ge(Box<Comparison>, Term),
    Gt(Box<Comparison>, Term),
    Le(Box<Comparison>, Term),
    Lt(Box<Comparison>, Term),
}

impl Comparison {
    pub fn self_terminating(&self) -> bool {
        match self {
            Comparison::Value(term) => term.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Term {
    Value(Factor),

    Add(Box<Term>, Factor),
    Sub(Box<Term>, Factor),
}

impl Term {
    pub fn self_terminating(&self) -> bool {
        match self {
            Term::Value(factor) => factor.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Factor {
    Value(Unary),

    Mul(Box<Factor>, Unary),
    Div(Box<Factor>, Unary),
    Rem(Box<Factor>, Unary),
}

impl Factor {
    pub fn self_terminating(&self) -> bool {
        match self {
            Factor::Value(unary) => unary.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Unary {
    Value(Call),

    Neg(Box<Unary>),
    Not(Box<Unary>),
}

impl Unary {
    pub fn self_terminating(&self) -> bool {
        match self {
            Unary::Value(call) => call.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Call {
    Value(Primary),

    Call(Box<Call>, Vec<Expression>),
    Index(Box<Call>, Box<Expression>),
    Field(Box<Call>, Identifier),
}

impl Call {
    pub fn self_terminating(&self) -> bool {
        match self {
            Call::Value(call) => call.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Primary {
    Block(Block),
    If(If),
    Group(Box<Expression>),

    Atom(Atom),
}

#[derive(Debug, Clone, Serialize)]
pub struct Object {
    pub ty: Identifier,
    pub fields: Vec<(Identifier, Expression)>,
}

#[derive(Debug, Clone, Serialize)]
pub struct List {
    pub items: Vec<Expression>,
}

impl Primary {
    pub fn self_terminating(&self) -> bool {
        match self {
            Primary::Block(_) | Primary::If(_) => true,
            Primary::Atom(_) | Primary::Group(_) => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Atom {
    Identifier(Identifier),
    Literal(Literal),
    Object(Object),
    List(List),
}

#[derive(Debug, Clone, Serialize)]
pub struct Assignment {
    pub place: Place,
    pub expr: Expression,
}

#[derive(Debug, Clone, Serialize)]
pub enum Place {
    Field(Call, Identifier),
    Identifier(Identifier),
    Index(Call, Box<Expression>),
}

#[derive(thiserror::Error, Debug, Clone, Serialize)]
#[error("cannot assign to `{:?}`", target)]
pub struct PlaceError {
    pub target: Expression,
}

impl TryFrom<Expression> for Place {
    type Error = PlaceError;

    fn try_from(expr: Expression) -> Result<Self, Self::Error> {
        let target = expr.clone();

        let Expression::LogicOr(expr) = expr;

        let LogicOr { first: expr, rest } = expr;
        if !rest.is_empty() {
            return Err(PlaceError { target });
        }

        let LogicAnd { first: expr, rest } = &expr;
        if !rest.is_empty() {
            return Err(PlaceError { target });
        }

        let Equality::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Comparison::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Term::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Factor::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Unary::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        match expr {
            Call::Value(expr) => {
                let Primary::Atom(expr) = expr else {
                    return Err(PlaceError { target });
                };

                let Atom::Identifier(ident) = expr else {
                    return Err(PlaceError { target });
                };

                // Prevent assigning to `self` to avoid implying that the receiver can be changed
                // that way.
                if ident.name == "self" {
                    return Err(PlaceError { target });
                }

                Ok(Place::Identifier(ident.clone()))
            }
            Call::Field(obj, ident) => Ok(Place::Field(*obj.clone(), ident.clone())),
            Call::Index(obj, key) => Ok(Place::Index(*obj.clone(), key.clone())),
            Call::Call(_, _) => Err(PlaceError { target }),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct Type {
    pub ident: Identifier,
    pub methods: Vec<Method>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Method {
    pub self_: bool,
    pub ident: Identifier,
    pub params: Vec<Identifier>,
    pub body: Block,
}

#[derive(Debug, Clone, Serialize)]
pub struct Func {
    pub ident: Identifier,
    pub params: Vec<Identifier>,
    pub body: Block,
}

#[derive(Debug, Clone, Serialize)]
pub struct Let {
    pub ident: Identifier,
    pub expr: Expression,
}

#[derive(Debug, Clone, Serialize)]
pub struct Loop {
    pub label: Option<Identifier>,
    pub body: Block,
}

#[derive(Debug, Clone, Serialize)]
pub struct Return {
    pub expr: Expression,
}

#[derive(Debug, Clone, Serialize)]
pub struct Break {
    pub label: Option<Identifier>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Continue {
    pub label: Option<Identifier>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Block {
    pub decls: Vec<Declaration>,
    pub expr: Option<Box<Expression>>,
}

#[derive(Debug, Clone, Serialize)]
pub struct If {
    pub condition: Box<Expression>,
    pub consequent: Block,
    pub alternative: Option<Block>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Identifier {
    pub name: String,
}

impl Identifier {
    pub fn new(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }

    pub fn empty() -> Self {
        Self {
            name: String::from(""),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Literal {
    Nil,
    Boolean(bool),
    Integer(BigInt),
    Float(f64),
    String(String),
}
