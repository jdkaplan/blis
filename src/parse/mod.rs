pub mod ast;
pub mod lexer;
pub mod parser;
pub mod token;

pub use lexer::Lexer;
pub use parser::{ParseError, ParseErrors, Parser};
pub use token::{Lexeme, LexemeOwned, Token};
