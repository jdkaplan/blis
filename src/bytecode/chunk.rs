use std::io;

use num_rational::BigRational;
use serde::{Deserialize, Serialize};

use crate::bytecode::disassembly::DisassembledChunk;
use crate::bytecode::{Func, Op, OpError};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, strum::EnumIs, strum::EnumTryAs)]
pub enum Constant {
    Rational(BigRational),
    Float(f64),
    String(String),
    Func(Func),
}

#[derive(Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Chunk {
    pub constants: Vec<Constant>,
    pub globals: Vec<String>,
    pub code: Vec<u8>,
}

impl std::fmt::Debug for Chunk {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Chunk")
            .field("constants", &format!("(count: {})", self.constants.len()))
            .field("globals", &format!("(count: {})", self.globals.len()))
            .field("code", &format!("(bytes: {})", self.code.len()))
            .finish()
    }
}

// Code
impl Chunk {
    pub fn push(&mut self, op: Op) {
        self.push_bytes(&op.to_bytes())
    }

    pub fn push_bytes(&mut self, bytes: &[u8]) {
        self.code.extend(bytes)
    }

    pub fn add_constant(&mut self, constant: Constant) -> u8 {
        // Reuse existing constants when possible.
        if let Some(idx) = self.find_constant(&constant) {
            return idx;
        }

        let idx = self.constants.len();
        assert!(idx < u8::MAX.into());

        self.constants.push(constant);
        idx as u8
    }

    fn find_constant(&self, constant: &Constant) -> Option<u8> {
        // TODO: HashMap/BTreeMap?
        self.constants
            .iter()
            .position(|c| c == constant)
            .map(|idx| idx.try_into().expect("add_constant enforces count"))
    }
}

// Globals
impl Chunk {
    pub fn define_global(&mut self, name: String) {
        let id = self.make_global(name);
        self.push(Op::DefineGlobal(id));
    }

    fn find_global(&self, name: &str) -> Option<u8> {
        // TODO: HashMap/BTreeMap?
        self.globals
            .iter()
            .position(|g| g == name)
            .map(|idx| idx.try_into().expect("make_global enforces count"))
    }

    pub fn make_global(&mut self, name: String) -> u8 {
        if let Some(id) = self.find_global(&name) {
            return id;
        }

        let idx = self.globals.len();
        let idx: u8 = idx
            .try_into()
            .expect("less than 256 constants, needs GlobalDefine2?");

        self.globals.push(name);
        idx
    }
}

#[must_use]
#[derive(Debug, Copy, Clone)]
pub struct PendingJump(usize);

// Jumps
impl Chunk {
    #[must_use = "set_jump_target"]
    pub fn prepare_jump(&mut self, op: Op) -> PendingJump {
        let idx = self.code.len();
        self.push(op);
        PendingJump(idx)
    }

    pub fn set_jump_target(&mut self, jump: PendingJump) {
        let idx = jump.0;
        let target = self.code.len();

        let offset: u16 = target
            .checked_sub(idx)
            .expect("jump is always backwards")
            .try_into()
            .expect("jump offset fits in two bytes");

        let [hi, lo] = offset.to_be_bytes();
        self.code[idx + 1] = hi;
        self.code[idx + 2] = lo;
    }
}

#[derive(thiserror::Error, Debug)]
pub enum ChunkReadError {
    #[error(transparent)]
    Deserialize(postcard::Error),

    #[error("extra bytes at end of file: {0:?}")]
    ExtraBytes(Vec<u8>),
}

#[derive(thiserror::Error, Debug)]
pub enum ChunkWriteError {
    #[error(transparent)]
    Serialize(postcard::Error),
}

// Compiled files
impl Chunk {
    pub fn read(r: impl io::Read) -> Result<Self, ChunkReadError> {
        let mut extra_bytes = Vec::new();

        let (chunk, (_, _)) =
            postcard::from_io((r, &mut extra_bytes)).map_err(ChunkReadError::Deserialize)?;

        if extra_bytes.is_empty() {
            Ok(chunk)
        } else {
            Err(ChunkReadError::ExtraBytes(extra_bytes))
        }
    }

    pub fn write(&self, w: impl io::Write) -> Result<(), ChunkWriteError> {
        postcard::to_io(self, w).map_err(ChunkWriteError::Serialize)?;
        Ok(())
    }
}

impl Chunk {
    pub fn iter_code(&self) -> CodeIterator<'_> {
        CodeIterator {
            code: &self.code,
            pc: 0,
            errored: false,
        }
    }
}

pub struct CodeIterator<'a> {
    code: &'a [u8],
    pc: usize,
    errored: bool,
}

impl Iterator for CodeIterator<'_> {
    type Item = Result<(usize, Op), OpError>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.errored {
            return None;
        }

        let pc = self.pc;
        match Op::scan(&self.code[pc..]) {
            Ok(None) => None,

            Ok(Some(op)) => {
                self.pc += op.size_bytes();
                Some(Ok((pc, op)))
            }

            Err(err) => {
                self.errored = true;
                Some(Err(err))
            }
        }
    }
}

impl Chunk {
    pub fn disassemble(&self) -> DisassembledChunk {
        self.into()
    }
}
