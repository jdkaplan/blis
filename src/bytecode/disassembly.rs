use num_rational::BigRational;
use serde::{Deserialize, Serialize};

use crate::bytecode::{Chunk, Constant, Func, Op};

#[derive(Clone, Serialize, Deserialize)]
pub struct DisassembledChunk {
    pub constants: Vec<DisassembledConstant>,
    pub globals: Vec<String>,
    pub code: Vec<Op>,
}

impl From<&Chunk> for DisassembledChunk {
    fn from(chunk: &Chunk) -> Self {
        let constants = chunk.constants.iter().map(Into::into).collect();
        let code: Result<Vec<_>, _> = chunk
            .iter_code()
            .map(|res| res.map(|(_pc, op)| op))
            .collect();

        Self {
            constants,
            code: code.unwrap(),
            globals: chunk.globals.clone(),
        }
    }
}

#[derive(Clone, Serialize, Deserialize, strum::EnumIs, strum::EnumTryAs)]
pub enum DisassembledConstant {
    Rational(BigRational),
    Float(f64),
    String(String),
    Func(DisassembledFunc),
}

impl From<&Constant> for DisassembledConstant {
    fn from(constant: &Constant) -> Self {
        match constant {
            Constant::Rational(v) => Self::Rational(v.clone()),
            Constant::Float(v) => Self::Float(*v),
            Constant::String(v) => Self::String(v.clone()),
            Constant::Func(func) => Self::Func(func.into()),
        }
    }
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DisassembledFunc {
    pub name: String,
    pub arity: u8,
    pub upvalues: u8,
    pub chunk: DisassembledChunk,
}

impl From<&Func> for DisassembledFunc {
    fn from(func: &Func) -> Self {
        Self {
            name: func.name.clone(),
            arity: func.arity,
            upvalues: func.upvalues,
            chunk: func.chunk.disassemble(),
        }
    }
}
