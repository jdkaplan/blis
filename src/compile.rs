use num_rational::BigRational;
use tracing::{debug, instrument};

use crate::bytecode::{Chunk, Constant, Op};
use crate::parse::ast::*;

#[derive(Debug)]
pub struct Compiler {
    chunk: Chunk,
    errors: Vec<CompileError>,

    locals: Locals,
}

#[derive(thiserror::Error, Debug)]
pub struct CompileErrors(Vec<CompileError>);

impl std::fmt::Display for CompileErrors {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(err) = self.0.first() {
            write!(f, "{}", err)?;
        } else {
            return write!(f, "");
        }

        for err in &self.0[1..] {
            write!(f, "\n{}", err)?;
        }
        Ok(())
    }
}

#[derive(thiserror::Error, Debug)]
pub enum CompileError {
    #[error(
        "variable redeclared in the same scope: `{}`",
        .0.name
    )]
    DuplicateVariable(Identifier),

    #[error("{0}")]
    Other(String),
}

type CompileResult<T> = Result<T, CompileError>;

type Fallible<T> = Result<T, FailedCodegen>;

#[derive(Debug)]
struct FailedCodegen;

impl Compiler {
    fn new() -> Self {
        Self {
            chunk: Chunk::default(),
            errors: Vec::new(),
            locals: Locals::default(),
        }
    }

    fn error(&mut self, err: CompileError) -> FailedCodegen {
        debug!({ ?err }, "compile error");
        self.errors.push(err);
        FailedCodegen
    }

    #[instrument(level = "debug", ret)]
    pub fn compile(program: &Program) -> Result<Chunk, CompileErrors> {
        let mut compiler = Self::new();
        let _ = compiler.program(program); // Errors checked below

        if compiler.errors.is_empty() {
            for (idx, val) in compiler.chunk.constants.iter().enumerate() {
                debug!({ ?idx, ?val }, "constant");
            }

            for (idx, name) in compiler.chunk.globals.iter().enumerate() {
                debug!({ ?idx, ?name }, "global");
            }

            for res in compiler.chunk.iter_code() {
                let Ok((pc, op)) = res else {
                    debug!({ ?res }, "chunk error");
                    break;
                };
                debug!({ ?pc, ?op }, "code");
            }

            Ok(compiler.chunk)
        } else {
            Err(CompileErrors(compiler.errors))
        }
    }

    #[instrument(level = "trace")]
    fn program(&mut self, program: &Program) -> Fallible<()> {
        for decl in &program.decls {
            self.declaration(decl)?;
        }
        self.chunk.push(Op::Return);
        Ok(())
    }

    #[instrument(level = "trace")]
    fn block(&mut self, block: &Block) -> Fallible<()> {
        self.locals.begin_scope();

        for decl in &block.decls {
            self.declaration(decl)?;
        }

        if let Some(expr) = &block.expr {
            self.expression(expr)?;
        } else {
            self.chunk.push(Op::Nil);
        }

        self.locals.end_scope();
        Ok(())
    }

    #[instrument(level = "trace")]
    fn declaration(&mut self, decl: &Declaration) -> Fallible<()> {
        match decl {
            Declaration::Func(func) => self.decl_func(func),
            Declaration::Let(let_) => self.decl_let(let_),
            Declaration::Statement(stmt) => self.statement(stmt),
        }
    }

    #[instrument(level = "trace")]
    fn decl_func(&mut self, func: &Func) -> Fallible<()> {
        // - Declare the identifier in the current scope
        // - Mark it as initialized so it's usable within the func body
        // - Start a new scope containing all the param names
        // - Start a new chunk for the implementation.
        // - Compile the function into the chunk.
        // - Make a constant to hold the func header
        // - Load the constant and store it in the local/global name
        todo!("start here")
    }

    #[instrument(level = "trace")]
    fn decl_let(&mut self, let_: &Let) -> Fallible<()> {
        let slot = self
            .locals
            .reserve(let_.ident.clone())
            .map_err(|err| self.error(err))?;

        self.expression(&let_.expr)?;

        self.locals.mark_init(slot);
        Ok(())
    }

    #[instrument(level = "trace")]
    fn statement(&mut self, stmt: &Statement) -> Fallible<()> {
        match stmt {
            Statement::Assignment(assign) => self.stmt_assign(assign),

            Statement::Expression(expr) => {
                self.expression(expr)?;
                self.chunk.push(Op::Pop);
                Ok(())
            }
        }
    }

    #[instrument(level = "trace")]
    fn stmt_assign(&mut self, assign: &Assignment) -> Fallible<()> {
        let Place::Identifier(ident) = &assign.place;

        self.expression(&assign.expr)?;

        if let Some((slot, _local)) = self.locals.resolve(&ident.name) {
            self.chunk.push(Op::LocalSet(slot));
        } else {
            let global = self.chunk.make_global(ident.name.clone());
            self.chunk.push(Op::GlobalSet(global));
        }

        Ok(())
    }

    #[instrument(level = "trace")]
    fn expression(&mut self, expr: &Expression) -> Fallible<()> {
        let Expression::LogicOr(or) = expr;
        self.logic_or(or)
    }

    #[instrument(level = "trace")]
    fn logic_or(&mut self, or: &LogicOr) -> Fallible<()> {
        // This leaves the last value that was considered on top of the stack.

        // <a>
        self.logic_and(&or.first)?;

        let mut shorts = Vec::new();
        for and in &or.rest {
            // jump_true_peek 'short_circuit
            let short = self.chunk.prepare_jump(Op::JumpTruePeek(i16::MAX));
            shorts.push(short);

            // pop ; <a> was falsey
            self.chunk.push(Op::Pop);

            // <b>
            self.logic_and(and)?;
        }

        // 'short_circuit:
        for jump in shorts {
            self.chunk.set_jump_target(jump);
        }

        Ok(())
    }

    fn logic_and(&mut self, and: &LogicAnd) -> Fallible<()> {
        // This leaves the last value that was considered on top of the stack.

        // <a>
        self.equality(&and.first)?;

        let mut shorts = Vec::new();
        for eq in &and.rest {
            // jump_false_peek 'short_circuit
            let short = self.chunk.prepare_jump(Op::JumpFalsePeek(i16::MAX));
            shorts.push(short);

            // pop ; <a> was truthy
            self.chunk.push(Op::Pop);

            // <b>
            self.equality(eq)?;
        }

        // 'short_circuit:
        for jump in shorts {
            self.chunk.set_jump_target(jump);
        }

        Ok(())
    }

    fn equality(&mut self, eq: &Equality) -> Fallible<()> {
        match eq {
            Equality::Value(comp) => self.comparison(comp),

            Equality::Eq(a, b) => {
                self.equality(a)?;
                self.comparison(b)?;
                self.chunk.push(Op::Eq);
                Ok(())
            }

            Equality::Ne(a, b) => {
                self.equality(a)?;
                self.comparison(b)?;
                self.chunk.push(Op::Ne);
                Ok(())
            }
        }
    }

    fn comparison(&mut self, comp: &Comparison) -> Fallible<()> {
        match comp {
            Comparison::Value(term) => self.term(term),

            Comparison::Gt(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Gt);
                Ok(())
            }

            Comparison::Ge(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Ge);
                Ok(())
            }

            Comparison::Lt(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Lt);
                Ok(())
            }

            Comparison::Le(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Le);
                Ok(())
            }
        }
    }

    fn term(&mut self, term: &Term) -> Fallible<()> {
        match term {
            Term::Value(factor) => self.factor(factor),

            Term::Add(a, b) => {
                self.term(a)?;
                self.factor(b)?;
                self.chunk.push(Op::Add);
                Ok(())
            }

            Term::Sub(a, b) => {
                self.term(a)?;
                self.factor(b)?;
                self.chunk.push(Op::Sub);
                Ok(())
            }
        }
    }

    fn factor(&mut self, factor: &Factor) -> Fallible<()> {
        match factor {
            Factor::Value(unary) => self.unary(unary),

            Factor::Mul(a, b) => {
                self.factor(a)?;
                self.unary(b)?;
                self.chunk.push(Op::Mul);
                Ok(())
            }

            Factor::Div(a, b) => {
                self.factor(a)?;
                self.unary(b)?;
                self.chunk.push(Op::Div);
                Ok(())
            }

            Factor::Rem(a, b) => {
                self.factor(a)?;
                self.unary(b)?;
                self.chunk.push(Op::Rem);
                Ok(())
            }
        }
    }

    fn unary(&mut self, unary: &Unary) -> Fallible<()> {
        match unary {
            Unary::Value(call) => self.call(call),

            Unary::Not(a) => {
                self.unary(a)?;
                self.chunk.push(Op::Not);
                Ok(())
            }

            Unary::Neg(a) => {
                self.unary(a)?;
                self.chunk.push(Op::Neg);
                Ok(())
            }
        }
    }

    fn call(&mut self, call: &Call) -> Fallible<()> {
        match call {
            Call::Value(primary) => self.primary(primary),

            Call::Call(callee, args) => {
                self.call(callee)?;
                for arg in args {
                    self.expression(arg)?;
                }

                let arity: u8 = args.len().try_into().expect("at most 255 args");
                self.chunk.push(Op::Call(arity));
                Ok(())
            }

            Call::Index(obj, key) => {
                self.call(obj)?;
                self.expression(key)?;
                self.chunk.push(Op::Index);
                Ok(())
            }

            Call::Field(obj, field) => {
                self.call(obj)?;

                let name = Constant::String(field.name.clone());
                let id = self.chunk.add_constant(name);
                self.chunk.push(Op::Constant(id));
                Ok(())
            }
        }
    }

    #[instrument(level = "trace")]
    fn primary(&mut self, primary: &Primary) -> Fallible<()> {
        match primary {
            Primary::Block(block) => self.block(block),
            Primary::If(if_) => self.expr_if(if_),
            Primary::Atom(atom) => self.atom(atom),
        }
    }

    #[instrument(level = "trace")]
    fn atom(&mut self, atom: &Atom) -> Fallible<()> {
        match atom {
            Atom::Identifier(ident) => self.identifier(ident),
            Atom::Literal(lit) => self.literal(lit),
        }
    }

    #[instrument(level = "trace")]
    fn expr_if(&mut self, if_: &If) -> Fallible<()> {
        // <cond>
        self.expression(&if_.condition)?;

        // jump_false_pop 'skip_conseq [peek cond]
        let skip_conseq = self.chunk.prepare_jump(Op::JumpFalsePop(i16::MAX));

        // <consequent>
        self.block(&if_.consequent)?;

        let Some(alt) = &if_.alternative else {
            // 'skip_conseq:
            self.chunk.set_jump_target(skip_conseq);
            self.chunk.push(Op::Nil);
            return Ok(());
        };

        // jump 'skip_alt
        let skip_alt = self.chunk.prepare_jump(Op::Jump(i16::MAX));

        // 'skip_conseq:
        self.chunk.set_jump_target(skip_conseq);

        // <alternative>
        self.block(alt)?;

        // 'skip_alt:
        self.chunk.set_jump_target(skip_alt);

        Ok(())
    }

    #[instrument(level = "trace")]
    fn identifier(&mut self, ident: &Identifier) -> Fallible<()> {
        if let Some((slot, _local)) = self.locals.resolve(&ident.name) {
            self.chunk.push(Op::LocalGet(slot));
        } else {
            let global = self.chunk.make_global(ident.name.clone());
            self.chunk.push(Op::GlobalGet(global));
        }
        Ok(())
    }

    #[instrument(level = "trace")]
    fn literal(&mut self, lit: &Literal) -> Fallible<()> {
        match lit {
            Literal::Nil => {
                self.chunk.push(Op::Nil);
            }
            Literal::Boolean(b) => {
                if *b {
                    self.chunk.push(Op::True);
                } else {
                    self.chunk.push(Op::False);
                }
            }

            Literal::Integer(v) => {
                let r = BigRational::new(v.clone(), 1.into());
                let id = self.chunk.add_constant(Constant::Rational(r));
                self.chunk.push(Op::Constant(id));
            }
            Literal::Float(v) => {
                let id = self.chunk.add_constant(Constant::Float(*v));
                self.chunk.push(Op::Constant(id));
            }
            Literal::String(v) => {
                let id = self.chunk.add_constant(Constant::String(v.clone()));
                self.chunk.push(Op::Constant(id));
            }
        }

        Ok(())
    }
}

#[derive(Debug, Clone)]
struct Local {
    depth: usize,
    initialized: bool,

    ident: Identifier,
}

#[derive(Debug, Clone, Default)]
struct Locals {
    locals: Vec<Local>,
    scope_depth: usize,
}

impl Locals {
    #[instrument(level = "trace")]
    fn begin_scope(&mut self) {
        self.scope_depth += 1;
    }

    #[instrument(level = "trace")]
    fn end_scope(&mut self) -> usize {
        self.scope_depth -= 1;

        let mut count = 0;

        while let Some(local) = self.locals.last() {
            if local.depth > self.scope_depth {
                self.locals.pop();
                count += 1;
            } else {
                break;
            }
        }

        count
    }

    fn iter_slots(&self) -> impl Iterator<Item = (usize, &Local)> {
        self.locals.iter().enumerate().rev()
    }

    fn iter_stack(&self) -> impl Iterator<Item = &Local> {
        self.locals.iter().rev()
    }

    #[instrument(level = "trace")]
    fn resolve<'a>(&'a self, name: &str) -> Option<(u8, &'a Local)> {
        for (slot, var) in self.iter_slots() {
            if var.ident.name == name {
                let slot: u8 = slot.try_into().expect("Too many locals!");
                return Some((slot, var));
            }
        }
        None
    }

    #[instrument(level = "trace")]
    fn reserve(&mut self, ident: Identifier) -> CompileResult<u8> {
        if self.locals.len() >= (u8::MAX as usize) {
            panic!("Too many locals! Time to add Local2 variants");
        }

        let new = Local {
            ident,
            depth: self.scope_depth,
            initialized: false,
        };

        for old in self.iter_stack() {
            if old.ident.name == new.ident.name && old.depth == new.depth {
                return Err(CompileError::DuplicateVariable(new.ident));
            }
        }

        let slot: u8 = self
            .locals
            .len()
            .try_into()
            .expect("Too many locals! Time to add Local2 variants");
        self.locals.push(new);
        Ok(slot)
    }

    #[instrument(level = "trace")]
    fn mark_init(&mut self, slot: u8) {
        self.locals[slot as usize].initialized = true;
    }
}
