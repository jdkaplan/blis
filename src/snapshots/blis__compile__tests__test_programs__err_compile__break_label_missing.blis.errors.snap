---
source: src/compile.rs
description: "loop outer {\n    break inner;\n}\n"
input_file: test_programs/err_compile/break_label_missing.blis
---
CompileErrors([
  NoBreakLabel(Identifier(
    name: "inner",
  )),
])
